File: go/kan/coyoneda.go
package capitan

type Coyoneda struct {
    f func(interface{}) interface{}
    m interface{}
}

func NewCoyoneda(f func(interface{}) interface{}, m interface{}) *Coyoneda {
    return &Coyoneda{f, m}
}

func (c *Coyoneda) ContraMap(g func(interface{}) interface{}) *Coyoneda {
    return &Coyoneda{
        f: func(x interface{}) interface{} {
            return g(c.f(x))
        },
        m: c.m,
    }
}

File: python/kan/coyoneda.py
from typing import Callable, Generic, TypeVar

F = TypeVar('F')
A = TypeVar('A')
B = TypeVar('B')

class Coyoneda(Generic[F, A]):
    def __init__(self, f: Callable[[A], F], m: F):
        self.f = f
        self.m = m

    def contramap(self, g: Callable[[B], A]) -> 'Coyoneda[F, B]':
        return Coyoneda(lambda b: self.f(g(b)), self.m)